enhancements/network/baremetal-ipi-network-configuration.md

```diff
@@ -0,0 +1,187 @@
+
+The data provided by the user will need to have the following structure:
+```
+[hostname]:
+  [interface name]:
+    [contents of NetworkManager key file for interface]
+  [second interface name]:
+    [contents of NetworkManager key file for second interface]
+  etc...
+```
+
+For example:
+```
+master-0:
+  eth0:
+    [connection]
+    id=eth0
+    uuid=18e0cec7-041c-4fb3-957c-a60c80dd9b85
+    type=ethernet
+    etc...
+master-1:
+  eth0:
+    [connection]
+    id=eth0
+    etc...
+```
+
+#### API Open Questions
+
+* How do we map an individual machine's keyfiles to the machine? Hostname was previously suggested because it was already an available label in NNCP, but if we're not using NNCP then there might be a better way.
+
+* If we're using RHCOS functionality, do we still need the custom image build discussed below? Or can we just pass the keyfiles as part of the deployment process?
+
+* What is the best way to encode the data above into secret(s)?
+
+### Processing user configuration
+
+#### Deployment of the controlplane hosts via terraform
+
+TODO - we'll have to match the keyfiles to the master Machines, then generate image/images for deployment via terraform/ironic
+
+#### Deployment of compute/worker hosts
+
+TODO - this will rely on metal3 image-building proposal https://github.com/metal3-io/metal3-docs/pull/183
+
+## Alternatives
+
+### Use Kubernetes-NMState NodeNetworkConfigurationPolicy custom resources
+
+If we were able to install the NNCP CRD at day-1 then we could use that as the configuration interface. This has the advantage of matching the configuration syntax and objects used for day-2 network configuration via the operator.
+
+However, we have been unable to get agreement on a method to deliver the operator on day-1, which makes this problematic. It's possible we could implement some way to pull just the NNCP CRD from the operator payload and isntall it on day-1, but at this point we likely don't have time to do that and it's not clear we even want to if the operator can't be installed on day-1 too. This would essentially orphan the CRs after the initial deployment completes.
+
+### Provide raw nmstate data
+
+We could expose an interface to the user which accepts raw nmstate data (not a CR wrapper).  This would mean a common DSL to the NodeNetworkConfigurationPolicy resource, but not directly reusing that API, so there is risk of drift between the two APIs.
+
+The user would have to pass the nmstate data via the install-config, and also as a Secret for use on scale-up, then also define a NNCP resource if day-2 management of the configuration is required (which could be risky if the configuration isn't exactly the same...)
```

Comment:
I think that this problem, of having 3 different copies of the same thing at
* install-config
* secret
* day-2 NNCP

exists also in the proposed solution; it's not a special downside of this alternative.

Reply:
The install-config data doesn't exist after installation, but yes I agree - we'll potentially have two copies of network configuration which isn't ideal.

Until we get a fully supported kubernetes-nmstate and an API derived from that included by default in OCP, I don't see how this can be avoided though.  We had a [previous draft proposal](https://github.com/hardys/enhancements/blob/b6a348dc7951b792fc338b1d0a237b023f18ec88/enhancements/installer/network_configuration.md) which aimed to use NNCP for both day-1 and day-2 - but unless that API is available on day-1 (which seems unlikely given #747 is stalled) that doesn't seem to be a viable option.

Reply:
We agree about this. But it is wrong to attach this to this alternative; this text should move up.

Reply:
Yeah, this entry was largely referring to a design we've abandoned for now. I'll update it to compare/contrast with the current proposal. I think we've already discussed the drawbacks in terms of duplicate configuration in the Risks and Mitigations section.